cppwinrt/code_writers.h

```diff
@@ -1148,7 +1154,10 @@ namespace cppwinrt
         {
             format = R"(    template <typename D%> auto consume_%<D%>::%(%) const
     {%
-        check_hresult(WINRT_IMPL_SHIM(%)->%(%));%
+        const auto castedResult = static_cast<% const&>(static_cast<D const&>(*this));
+        const auto abiType = *(abi_t<%>**)&castedResult;
+        check_cast_result(abiType);
+        check_hresult(abiType->%(%));%
     }
```

Comment:
You may also want to confirm (manually) that this still destroys the temporary. Its very easy to inject extra QIs or drop needed dtors in crazy code like this. I speak from experience. 😉

Reply:
I honestly think the verbosity is a good thing, not a bad thing.  I've been debugging and bottomed out at the `WINRT_IMPL_SHIM` macro many times and it has always confused me.  I had to stare at it for a good 20 minutes yesterday, with help from @jonwis, and it's still difficult for me to parse.  I think splitting out the two sets of casts into separate lines outside a macro makes the code much more explicit/debuggable for mortals.  It also makes space to set breakpoints.  I don't think that conciseness of generated code that humans don't edit is a top concern (within reason).

As for the temporary, what do you think the best way would be to check that?  This is keeping the winrt projection object alive through the end of the function so the lifetime should be fine (including destruction).  I had to expand this out because previously the temporary didn't outlive the macro so the calls after `WINRT_IMPL_SHIM` were using freed memory.

Reply:
I'd just want to confirm that:
* you're not getting a `QueryInterface` in cases where one isn't needed
* you're getting the right number of `Release` calls

The former probably requires a breakpoint or `printf` inside the `QueryInterface` and the latter can (cheaply) be stress-tested in the scratch project with a loop and observing that memory isn't being leaked.

Reply:
I added a little test program to the scratch project.  When it is waiting for input before exiting the memory usage is very low (1MB active memory in task manager; close to the lowest on my system).  I think that confirms there is no leak.

I set breakpoints in QI, AddRef, and Release.  When this loop is running there is one AddRef for the initial cast.  There is one Release when the consume_ method returns.  This matches expectations.  There is no QI call in this case because it is already the correct type.

```cpp
namespace
{
    struct BasicStringable : public winrt::implements<BasicStringable, winrt::Windows::Foundation::IStringable>
    {
        BasicStringable() = default;
        hstring ToString()
        {
            return L"A string!";
        }
    };
}

int main()
{
    winrt::init_apartment();

    auto stringable = winrt::make_self<BasicStringable>().as<winrt::Windows::Foundation::IStringable>();
    for (auto i = 0U; i < 1000000; ++i)
    {
        const auto result = stringable.ToString();
    }

    printf("Press any key to exit");
    int num{};
    scanf("%d", &num);
}
```

Reply:
For the memory leak check, you'd want to move the "make" into the loop to put all the pressure on allocation rather than only reference counting. For example:

```C++
using namespace winrt;
using namespace Windows::Foundation;

struct Thing : implements<Thing, IClosable>
{
    void Close() {}
};

int main() {
    for (;;)
    {
        auto thing = make<Thing>();
        //detach_abi(thing);
    }
}
```

Then just run in the debugger and look for this flat graph:

![image](https://github.com/user-attachments/assets/20c27759-a5b7-4302-bff5-acca736eea65)

If i looks more like this then you have a problem: 

![image](https://github.com/user-attachments/assets/cc8195dd-91e9-4652-a9c4-1351264c4c72)

You can uncomment the `detach_abi` call to simulate.

Reply:
It is flat :)

![image](https://github.com/user-attachments/assets/900a1813-f4b9-4ee2-b914-bb97a3d248ec)

That is with the make_self call moved inside the loop.  I have not been able to get the scratch project to build projections from an IDL file so I can't easily test the case where the consume_ call requires a QI.  I can't get the scratch project to build an IDL at all (I've been trying for about an hour).

Reply:
Just rebuild the test projection with `build_projection.cmd` and then use a stock implementation like `Uri`.

Reply:
I realized after entirely too long that I don't need the scratch project for this.  I already have a little throwaway console app that builds just fine.  I made the changes there.

I'm observing that for the case where there is a QI in this code path there is a spurious reference.  The QI adds one reference.  And then a second addref is also triggered.

![image](https://github.com/user-attachments/assets/93015738-ab86-4a68-9782-74d88f6b0adc)

The proper number of Release calls are made so this does not leak, at least.  (I'll need to compare against the baseline to see if this is new from my changes or not).

Reply:
Alas, the baseline does not have that extra ref:

![image](https://github.com/user-attachments/assets/698a42c3-35ed-4e73-ae58-d6d58b06d9cd)

Reply:
That was building unoptimized.  When I build release everything is aggressively inlined into the containing code and there does not seem to be extra references:

![image](https://github.com/user-attachments/assets/feac3d06-2409-4b9e-8c65-f254b5269a13)

Reply:
It should definitely be unchanged in unoptimized builds.

Reply:
Thanks @DefaultRyan for pointing out that making castedResult a reference will fix the extra copy.  I have confirmed that in Debug builds the extra copy is gone.  And for Release there is no leak.

Debug:
![image](https://github.com/user-attachments/assets/08012c91-f23a-49b8-8c1e-86ff286ef296)

Release:
![image](https://github.com/user-attachments/assets/69ff6f27-f448-4eaf-a2f0-e05832385ff8)

Reply:
Marking this thread as resolved.  I think the code gen is now confirmed as acceptable.  LMK if you disagree.